### PR TITLE
BoxView/BoxEdit: implement Error Handling

### PR DESCRIPTION
**Pull request by vahidbazzaz**
In the PR, error/success notification messages are handled consistently using a unified handler as follows:

- A local variable is created using the Apollo client
- Include a unify message handler in the layout
- Add a title to the notification message component
- update the error/success messages in the box edit and box view

### CODE REVIEW

**Comment by vahidbazzaz** [comment]
[2.0 BoxView/BoxEdit: implement Error Handling](https://trello.com/c/nEPzsu8F)

File: react/src/views/Box/BoxView.tsx [hunk 1]
```diff
@@ -283,7 +282,21 @@ function BTBox() {
...
+          title: `Box ${labelIdentifier}`,
+          type: "success",
+          message: `Successfully updated the box state to ${newState} `,
+        });
+      })
+      .catch(() => {
+        notificationVar({
+          title: `Box ${labelIdentifier}`,
+          type: "error",
+          message: "Error while trying to updat the Box state",
```
Comment (pylipp) [inline react/src/views/Box/BoxView.tsx:297]: Small typo: updat*e*. Also *b*ox instead of Box

File: react/src/views/Box/BoxView.tsx [hunk 2]
```diff
@@ -317,11 +338,19 @@ function BTBox() {
         },
       })
         .then(() => {
+          notificationVar({
+            title: `Box ${boxData.labelIdentifier}`,
+            type: "success",
+            message: `Successfully add ${boxFormValues?.numberOfItems}x to the box`,
```
Comment (pylipp) [inline react/src/views/Box/BoxView.tsx:344]: Similar to above: `Successfully added ${...} items to box`

File: react/src/views/Box/BoxView.tsx [hunk 3]
```diff
@@ -333,7 +362,21 @@ function BTBox() {
         newLocationId: parseInt(locationId, 10),
       },
       refetchQueries: [refetchBoxByLabelIdentifierQueryConfig(labelIdentifier)],
```
Comment (pylipp) [inline react/src/views/Box/BoxView.tsx:364]: It's not related to the direct changes of this PR but I was wondering why we refetch here. It's additional overhead whereas the new location could already be returned as response from the `updateBoxLocation` in the first place.

File: react/src/views/BoxEdit/components/BoxEdit.tsx [hunk 4]
```diff
@@ -195,14 +196,21 @@ function BoxEdit({
   }, [productId, productAndSizesData, boxData, resetField]);
 
   if (boxData == null) {
-    return <Box>No data found for a box with this id</Box>;
+    notificationVar({
```
Comment (pylipp) [inline react/src/views/BoxEdit/components/BoxEdit.tsx:199]: Does it make sense to have this check right away at the start of the function in line 108 or am I missing something?

File: react/src/views/Box/BoxView.tsx (hunk 3 above)
Comment (vahidbazzaz) [reply react/src/views/Box/BoxView.tsx:364]: @pylipp You are completely right, however, trying to remove the refetch did not update the box state anymore, so I need to investigate further why it was originally added.

Comment (pylipp) [reply react/src/views/Box/BoxView.tsx:370]: Thanks! I don't think it's super pressing but probably not an ideal solution either.

**Review by HaGuesto — COMMENTED** [review]
I screened through it and well done @vahidbazzaz!

I would have two questions:
1. How would Roanna test the errors on staging (or at least how should we test them?)?
2. I'm missing error logging to sentry. I'm not sure here though. Maybe a topic to discuss with @pylipp 
I believe graphql errors should be logged on the BE side, not sure if we also have to do that on FE. All other errors on FE should be logged when an error is caught.

**Comment by vahidbazzaz** [comment]
> I screened through it and well done @vahidbazzaz!
> 
> I would have two questions:
> 
> 1. How would Roanna test the errors on staging (or at least how should we test them?)?
> 2. I'm missing error logging to sentry. I'm not sure here though. Maybe a topic to discuss with @pylipp
>    I believe graphql errors should be logged on the BE side, not sure if we also have to do that on FE. All other errors on FE should be logged when an error is caught.

@HaGuesto Thank you for checking the PR. As we discussed, test cases must be written for the errors, and Sentry in the FE must be configured to log necessary errors.

**Review by HaGuesto — APPROVED** [review]

File: react/src/views/Box/BoxView.tsx [hunk 5]
```diff
@@ -283,7 +282,23 @@ function BTBox() {
...
       },
       // refetchQueries: [refetchBoxByLabelIdentifierQueryConfig(labelIdentifier)],
-    });
+    })
+      .then((res) => {
+        notificationVar({
+          title: `Box ${labelIdentifier}`,
+          type: res?.errors ? "error" : "success",
+          message: res?.errors
+            ? "Error while trying to update the box state"
```
Comment (aerinsol) [inline react/src/views/Box/BoxView.tsx:291]: "Error: Could not update the box state to {state}." should be the error. Then what happens? Does the whole action fail? Should the user try again?

File: react/src/views/Box/BoxView.tsx [hunk 6]
```diff
@@ -294,12 +309,22 @@ function BTBox() {
...
           onMinusClose();
         })
-        .catch((e) => {
-          // eslint-disable-next-line no-console
-          console.error("Error while trying to change number of items in the Box", e);
+        .catch(() => {
+          notificationVar({
+            title: `Box ${boxData.labelIdentifier}`,
+            type: "error",
+            message: "Error while trying to change number of items in the box",
```
Comment (aerinsol) [inline react/src/views/Box/BoxView.tsx:326]: "Error: Could not add/remove items from the box." should be the error. Then what happens? Does the whole action fail? Should the user try again?

File: react/src/views/Box/BoxView.tsx [hunk 7]
```diff
@@ -316,12 +341,22 @@ function BTBox() {
           numberOfItems: (boxData?.numberOfItems || 0) + (boxFormValues?.numberOfItems || 0),
         },
       })
-        .then(() => {
+        .then((res) => {
+          notificationVar({
+            title: `Box ${boxData.labelIdentifier}`,
+            type: res.errors ? "error" : "success",
+            message: res.errors
+              ? "Error while trying to change number of items in the box"
```
Comment (aerinsol) [inline react/src/views/Box/BoxView.tsx:349]: "Error: Could not add/remove items from the box." should be the error. Then what happens? Does the whole action fail? Should the user try again?

File: react/src/views/Box/BoxView.tsx [hunk 8]
```diff
@@ -316,12 +341,22 @@ function BTBox() {
...
           onPlusClose();
         })
-        .catch((e) => {
-          // eslint-disable-next-line no-console
-          console.error("Error while trying to change number of items in the Box", e);
+        .catch(() => {
+          notificationVar({
+            title: `Box ${boxData.labelIdentifier}`,
+            type: "error",
+            message: "Error while trying to change number of items in the box",
```
Comment (aerinsol) [inline react/src/views/Box/BoxView.tsx:358]: "Error: Could not add/remove items from the box" should be the error. Then what happens? Does the whole action fail? Should the user try again?

File: react/src/views/Box/BoxView.tsx [hunk 9]
```diff
@@ -333,7 +368,21 @@ function BTBox() {
         newLocationId: parseInt(locationId, 10),
       },
       refetchQueries: [refetchBoxByLabelIdentifierQueryConfig(labelIdentifier)],
-    });
+    })
+      .then((res) => {
+        notificationVar({
+          title: `Box ${labelIdentifier}`,
+          type: res.errors ? "error" : "success",
+          message: res.errors ? "Error while trying to move the box" : "Successfully moved the box",
```
Comment (aerinsol) [inline react/src/views/Box/BoxView.tsx:376]: "Error: Box could not be moved!" should be the error. Then what happens? Does the whole action fail? Should the user try again?

File: react/src/views/Box/BoxView.tsx [hunk 10]
```diff
@@ -333,7 +368,21 @@ function BTBox() {
...
+        });
+      })
+      .catch(() => {
+        notificationVar({
+          title: `Box ${labelIdentifier}`,
+          type: "error",
+          message: "Error while trying to move the box",
```
Comment (aerinsol) [inline react/src/views/Box/BoxView.tsx:383]: "Error: Box could not be moved!" should be the error. Then what happens? Does the whole action fail? Should the user try again?

File: react/src/views/BoxEdit/BoxEditView.tsx [hunk 11]
```diff
@@ -141,11 +142,36 @@ function BoxEditView() {
       },
     })
       .then((mutationResult) => {
-        navigate(`/bases/${baseId}/boxes/${mutationResult.data?.updateBox?.labelIdentifier}`);
+        if (mutationResult?.errors) {
+          notificationVar({
+            title: `Box ${labelIdentifier}`,
+            type: "error",
+            message: "Error while trying to update Box",
```
Comment (aerinsol) [inline react/src/views/BoxEdit/BoxEditView.tsx:149]: why does the error on line 173 have an error code, and this does not?

File: react/src/views/BoxEdit/BoxEditView.tsx [hunk 12]
```diff
@@ -141,11 +142,36 @@ function BoxEditView() {
...
+          });
+        } else {
+          notificationVar({
+            title: `Box ${labelIdentifier}`,
+            type: "success",
+            message: `Successfully modified with ${
```
Comment (vahidbazzaz) [inline react/src/views/BoxEdit/BoxEditView.tsx:155]: @aerinsol success message after box update: e.g. Successfully modified with baby towels Unisex Baby (75x) in LOST

File: react/src/views/BoxEdit/BoxEditView.tsx [hunk 13]
```diff
@@ -141,11 +142,36 @@ function BoxEditView() {
...
+          navigate(`/bases/${baseId}/boxes/${mutationResult.data?.updateBox?.labelIdentifier}`);
+        }
       })
       .catch((error) => {
-        // eslint-disable-next-line no-console
-        console.error("Error while trying to update Box", error);
+        notificationVar({
+          title: `Box ${labelIdentifier}`,
+          type: "error",
+          message: `Error while trying to update Box (code: ${error.code})`,
```
Comment (aerinsol) [inline react/src/views/BoxEdit/BoxEditView.tsx:173]: "Error - Code ${error.code}: Your changes could not be saved!"  Here the user should definitely stay on the page in order to not lose data. However, what's the expected behavior? Should the user try again?

File: react/src/views/BoxEdit/BoxEditView.tsx [hunk 14]
```diff
@@ -169,11 +195,21 @@ function BoxEditView() {
     }));
 
   if (allLocations == null) {
-    return <div>Error: no locations available to choose from</div>;
+    notificationVar({
+      title: "Error",
+      type: "error",
+      message: "No locations available to choose from",
```
Comment (aerinsol) [inline react/src/views/BoxEdit/BoxEditView.tsx:201]: "Error: No other locations are visible!"

File: react/src/views/BoxEdit/BoxEditView.tsx [hunk 15]
```diff
@@ -169,11 +195,21 @@ function BoxEditView() {
...
+    });
+    return <div />;
   }
 
   if (productAndSizesData?.elements == null) {
-    return <div>Error: no products available to choose from for this Box</div>;
+    notificationVar({
+      title: "Error",
+      type: "error",
+      message: "No products available to choose from for this Box",
```
Comment (aerinsol) [inline react/src/views/BoxEdit/BoxEditView.tsx:210]: "Error: No products are visible!"

File: react/src/views/BoxEdit/components/BoxEdit.tsx [hunk 16]
```diff
@@ -195,14 +196,21 @@ function BoxEdit({
...
+      title: "Error",
+      type: "error",
+      message: "No data found for a box with this id",
```
Comment (aerinsol) [inline react/src/views/BoxEdit/components/BoxEdit.tsx:202]: "Error: There is no data found for box {id}!"

File: react/src/views/BoxEdit/components/BoxEdit.tsx [hunk 17]
```diff
@@ -195,14 +196,21 @@ function BoxEdit({
...
 
   if (productsForDropdownGroups == null) {
-    // eslint-disable-next-line max-len
-    return (
-      <Box>There was an error: the available products to choose from couldn&apos;t be loaded!</Box>
-    );
+    notificationVar({
+      title: "Error",
+      type: "error",
+      message: "There was an error: the available products to choose from couldn&apos;t be loaded!",
```
Comment (aerinsol) [inline react/src/views/BoxEdit/components/BoxEdit.tsx:211]: "Error: The available products could not be loaded!"

**Review by aerinsol — CHANGES_REQUESTED** [review]
In general, I think I need to understand which errors are there because we want the user to try again, and which do they want to talk to us about /log in sentry.

File: react/src/views/Box/BoxView.tsx (hunk 5 above)
Comment (vahidbazzaz) [reply react/src/views/Box/BoxView.tsx:291]: @aerinsol yes the whole action fails

File: react/src/views/Box/BoxView.tsx (hunk 6 above)
Comment (vahidbazzaz) [reply react/src/views/Box/BoxView.tsx:326]: @aerinsol the same here, the whole action fail, and use should try again.

File: react/src/views/Box/BoxView.tsx (hunk 7 above)
Comment (vahidbazzaz) [reply react/src/views/Box/BoxView.tsx:349]: @aerinsol also here the same, the whole action fail and then user should try again

File: react/src/views/Box/BoxView.tsx (hunk 8 above)
Comment (vahidbazzaz) [reply react/src/views/Box/BoxView.tsx:358]: @aerinsol also here the same, the whole action fail and then user should try again

File: react/src/views/Box/BoxView.tsx (hunk 9 above)
Comment (vahidbazzaz) [reply react/src/views/Box/BoxView.tsx:376]: @aerinsol also here the same, the whole action fail and then user should try again

File: react/src/views/BoxEdit/BoxEditView.tsx (hunk 13 above)
Comment (vahidbazzaz) [reply react/src/views/BoxEdit/BoxEditView.tsx:173]: @aerinsol also here the same, the whole action fail and then user should try again

**Review by aerinsol — APPROVED** [review]
In general, I would suggest down the line we update the error messages with a call to action for what's needed from the user. However, since we haven't fully defined this on the FE yet, this is not a blocker. Approved.

**Pull request merged**